apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -265,80 +254,85 @@ const DanceAiModal: React.FunctionComponent<DanceAiModalProps> = ({
       [FieldKey.BACKGROUND_EFFECT]: backgroundLabels,
       [FieldKey.BACKGROUND_PALETTE]: paletteLabels,
     };
-  };
-
-  const getAllItems = () => {
-    return inputLibraryJson.items.map(item => {
-      return {
-        ...item,
-        available: !inputs.includes(item.id),
-      };
-    });
-  };
-
-  const getItem = (id: string) =>
-    inputLibraryJson.items.find(item => item.id === id);
+  }, []);
 
-  const handleItemClick = (id: string, available: boolean) => {
-    if (available) {
-      // Add item inputs.
-      if (currentInputSlot < SLOT_COUNT) {
-        setInputs([...inputs, id]);
-        setCurrentInputSlot(currentInputSlot + 1);
-        setInputAddCount(inputAddCount + 1);
+  const handleItemClick = useCallback(
+    (id: string, available: boolean) => {
+      setInputs(inputs => {
+        // if we've clicked on an emoji that's not available, remove it from our inputs
+        if (!available) {
+          playSound('ai-deselect-emoji');
+          return inputs.filter(input => input !== id);
+          // if we have more input slots, we can add the emoji
+        } else if (inputs.length < SLOT_COUNT) {
+          playSound('ai-select-emoji');
+          setInputAddCount(inputAddCount => inputAddCount + 1);
+          return [...inputs, id];
+          // otherwise, it's not a remove and we don't have space - do nothing.
+        } else {
+          return inputs;
+        }
+      });
+    },
+    [playSound]
+  );
 
-        playSound('ai-select-emoji');
-      }
-    } else {
-      // Remove item from inputs.
-      setInputs(inputs.filter(input => input !== id));
-      setCurrentInputSlot(currentInputSlot - 1);
+  const startAi = useCallback(() => {
+    generatedEffects.current = {
+      badEffects: getRangeArray(0, BAD_GENERATED_RESULTS_COUNT - 1).map(() =>
+        chooseEffects(inputs, ChooseEffectsQuality.BAD)
+      ),
+      goodEffect: chooseEffects(inputs, ChooseEffectsQuality.GOOD),
+    };
 
-      playSound('ai-deselect-emoji');
-    }
-  };
+    minMaxAssociations.current = calculateMinMax(inputs);
+  }, [calculateMinMax, inputs]);
 
-  const handleGenerateClick = () => {
+  const handleGenerateClick = useCallback(() => {
```

Comment:
All functions should be in `useCallback`, _especially_ event handlers, so as to prevent needless re-renders.